src/tango/quic/crypto/fd_quic_crypto_suites.c

```diff
@@ -811,3 +811,289 @@ fd_quic_crypto_rand( uchar * buf,
   ulong err = ERR_get_error();
   FD_LOG_ERR(( "openssl RAND_bytes failed. Error: %lu", err ));
 }
+
+int fd_quic_retry_token_encrypt(
+    fd_quic_conn_id_t * orig_dst_conn_id,
+    ulong *             now,
+    fd_quic_conn_id_t * retry_src_conn_id,
+    uint                ip_addr,
+    ushort              udp_port,
+    uchar               retry_token[static FD_QUIC_RETRY_TOKEN_CIPHERTEXT_SZ]
+) {
+  /* Generate pseudorandom bytes to use as the key for the AEAD HKDF. Note these bytes form the
+     beginning of the retry token. */
+  uchar * hkdf_key = retry_token;
+  int     rc       = fd_quic_crypto_rand( retry_token, FD_QUIC_RETRY_TOKEN_HKDF_KEY_SZ );
+  if ( FD_UNLIKELY( rc == FD_QUIC_FAILED ) ) {
+    return FD_QUIC_FAILED;
+  }
+
+  /* The `extract` step of HKDF is unnecessary because what's being passed in to `expand` are
+   pseudorandom bytes. */
+  uchar aead_key[FD_QUIC_RETRY_TOKEN_AEAD_KEY_SZ];
+  fd_quic_hkdf_expand_label(
+      aead_key,
+      FD_QUIC_RETRY_TOKEN_AEAD_KEY_SZ,
+      hkdf_key,
+      FD_QUIC_RETRY_TOKEN_HKDF_KEY_SZ,
+      FD_QUIC_RETRY_TOKEN_AEAD_INFO,
+      FD_QUIC_RETRY_TOKEN_AEAD_INFO_SZ,
+      fd_hmac_sha256,
+      FD_QUIC_RETRY_TOKEN_AEAD_KEY_SZ
+  );
+
+  /* Since the key is derived from random bytes and only used once, we use a zero IV (nonce).
+     Note the IV length is by default 12 bytes (which is the recommended length for AES-GCM). */
+  uchar iv[FD_QUIC_NONCE_SZ] = { 0 };
+
+  /* The AAD is the client IPv4 address, UDP port, and retry source connection id. */
+  ulong aad_sz = (ulong)FD_QUIC_RETRY_TOKEN_AAD_SZ + retry_src_conn_id->sz;
+  uchar aad[aad_sz];
+  memcpy( aad, &ip_addr, sizeof( uint ) );
+  memcpy( aad + sizeof( uint ), &udp_port, sizeof( ushort ) );
```

Comment:
the server is decoding the same bytes it encodes as-is. so the endianness is interpreted the same way it's written. otherwise the token is supposed to be treated as opaque by the client. good thing to check, thanks